src/client/src/services/layoutService.ts

```diff
@@ -47,3 +56,200 @@ export const dagreLayout = (elements: Elements, direction = 'TB'): Elements => {
 };
 
 //copy-paste ends
+//force-directed stuff
+
+interface coords {
+    x: number;
+    y: number;
+}
+
+//c1 - c2
+function subCoords(c1: coords, c2: coords): coords {
+    return { x: c1.x - c2.x, y: c1.y - c2.y };
+}
+
+function coordLength(c: coords) {
+    return Math.sqrt(c.x * c.x + c.y * c.y);
+}
+
+//moves c in the direction of dir for length amount
+function moveCoordToDir(c: coords, dir: coords, length: number) {
+    const dir_len = coordLength(dir);
+    c.x += (dir.x / dir_len) * length;
+    c.y += (dir.y / dir_len) * length;
+}
+
+function negCoord(c: coords) {
+    return { x: -c.x, y: -c.y };
+}
+
+function hasNaN(c: coords) {
+    return isNaN(c.x) || isNaN(c.y);
+}
+
+//does not scramble nodes, as this may also be used to make
+//the result of some analytic algorithm better
+export const forceDirectedLayout = (
+    elements: Elements,
+    iterations: number
+): Elements => {
+    const nodes: INode[] = [];
+    const edges: IEdge[] = [];
+
+    for (const el of elements) {
+        if (isNode(el)) nodes.push(el.data);
+        else if (isEdge(el))
+            edges.push({
+                source_id: el.source,
+                target_id: el.target,
+                project_id: 0,
+            });
+    }
+
+    const layoutedNodes = forceDirected(nodes, edges, iterations);
+
+    const nodeElements: Elements = layoutedNodes.map((n) => ({
+        id: String(n.id),
+        data: n,
+        position: { x: n.x, y: n.y },
+    }));
+
+    const edgeElements: Elements = edges.map((e) => ({
+        id: String(e.source_id) + '-' + String(e.target_id),
+        source: String(e.source_id),
+        target: String(e.target_id),
+        type: 'straight',
+        arrowHeadType: ArrowHeadType.ArrowClosed,
+        data: e,
+    }));
+
+    return nodeElements.concat(edgeElements);
+};
+
+const forceDirected = (nodes: INode[], edges: IEdge[], iterations: number) => {
+    const N = nodes.length;
+
+    //how much distance the nodes are allowed to move outwards of current graph
+    //TODO: maybe define somewhere else?
+    const margin = 100;
+
+    let x_min = 0;
+    let x_max = 10;
+    let y_min = 0;
+    let y_max = 10;
+
+    //node positions and edges will be transformed
+    //to a form that is more easily calculatable
+
+    const positions = Array<coords>(N);
+    const forces = Array<coords>(N); //not really forces but w/e
+    const id2idx = new Map<string, number>(); // node id -> index
+
+    //record node positions and calculate bounds
+    for (let i = 0; i < N; i++) {
+        const node = nodes[i];
+        positions[i] = { x: node.x, y: node.y };
+        //for some reason INode does not have to have id ???
+        if (node.id) id2idx.set(String(node.id) || 'wtf', i);
```

Comment:
This is because the DB is what creates the ID. And when you create a Node we use the same interface for both front and back end. When sending the node info to the server, it has all the info EXCEPT for the id